pkg/dockerregistry/server/maxconnections/maxconnections.go

```diff
@@ -0,0 +1,102 @@
+package maxconnections
+
+import (
+	"context"
+	"net/http"
+	"time"
+)
+
+func defaultOverloadHandler(w http.ResponseWriter, r *http.Request) {
+	http.Error(w, "503 service is overloaded, please try again later", http.StatusServiceUnavailable)
+}
+
+// DefaultOverloadHandler is a default OverloadHandler that used by New.
+var DefaultOverloadHandler http.Handler = http.HandlerFunc(defaultOverloadHandler)
+
+// Handler implements the http.Handler interface.
+type Handler struct {
+	// running is a buffered channel. Before invoking the handler, an empty
+	// struct is sent to the channel. When the handler is finished, one element
+	// is received back from the channel. If the channel's buffer is full, the
+	// request is enqueued.
+	running chan struct{}
+
+	// queue is a buffered channel. An empty struct is placed into the channel
+	// while a request is waiting for a spot in the running channel's buffer.
+	// If the queue channel's buffer is full, the request is proccess by
+	// OverloadHandler.
+	queue chan struct{}
+
+	// handler to invoke.
+	handler http.Handler
+
+	// MaxWaitInQueue is a maximum wait time in the queue.
+	MaxWaitInQueue time.Duration
+
+	// OverloadHandler is called if there is no space in running and queue
+	// channels.
+	OverloadHandler http.Handler
+
+	// newTimer allows to override the function newTimer for tests.
+	newTimer func(d time.Duration) *time.Timer
+}
+
+// New returns an http.Handler that runs no more than maxRunning h at the same
+// time. It can enqueue up to maxInQueue requests awaiting to be run, for other
+// requests OverloadHandler will be invoked.
+func New(maxRunning, maxInQueue int, h http.Handler) *Handler {
+	return &Handler{
+		running: make(chan struct{}, maxRunning),
+		queue:   make(chan struct{}, maxInQueue),
+		handler: h,
+
+		OverloadHandler: DefaultOverloadHandler,
+		newTimer:        time.NewTimer,
+	}
+}
+
+func (h *Handler) enqueueRunning(ctx context.Context) bool {
+	select {
+	case h.running <- struct{}{}:
+		return true
+	default:
+	}
+
+	// Slow-path.
+	select {
+	case h.queue <- struct{}{}:
+		defer func() {
+			<-h.queue
+		}()
+	default:
+		return false
+	}
+
+	var timer *time.Timer
+	var timeout <-chan time.Time
+	if h.MaxWaitInQueue > 0 {
```

Comment:
If MaxWaitInQueue is zero, why wouldn't we exit earlier in the function (i.e., before the enqueue above).